File: 0930-binary-subarrays-with-sum/0930-binary-subarrays-with-sum.py
class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        def getNumber(target):
            if target < 0: return 0

            lo, hi, cnt, total = 0, 0, 0, 0

            while hi < len(nums):
                total += nums[hi]
                while lo <= hi and total > target:
                    total -= nums[lo]
                    lo += 1
                cnt += hi - lo + 1
                hi += 1

            return cnt

        return getNumber(goal) - getNumber(goal - 1)

File: 0930-binary-subarrays-with-sum/0930-binary-subarrays-with-sum.java
class Solution {
    private int getNumber(int[] nums, int goal) {
        if (goal < 0) {
            return 0;
        }

        int lo = 0, hi = 0, cnt = 0, total = 0;

        while (hi < nums.length) {
            total += nums[hi];
            while (lo <= hi && total > goal) {
                total -= nums[lo];
                lo++;
            }
            cnt += hi - lo + 1;
            hi++;
        }

        return cnt;
    }

    public int numSubarraysWithSum(int[] nums, int goal) {
        return getNumber(nums, goal) - getNumber(nums, goal - 1);
    }
}